tests/test_default_headers.py

```diff
@@ -12,79 +12,69 @@ async def app(scope, receive, send):
     await send({"type": "http.response.body", "body": b"", "more_body": False})
 
 
-def test_default_default_headers():
-    config = Config(app=app, loop="asyncio", limit_max_requests=1)
-    server = Server(config=config)
-    thread = threading.Thread(target=server.run)
-    thread.start()
-    while not server.started:
-        time.sleep(0.01)
-    response = requests.get("http://127.0.0.1:8000")
-
-    assert response.headers["server"] == "uvicorn" and response.headers["date"]
-
-    thread.join()
-
-
-def test_override_server_header():
-    config = Config(
-        app=app,
-        loop="asyncio",
-        limit_max_requests=1,
-        headers=[("Server", "over-ridden")],
-    )
-    server = Server(config=config)
-    thread = threading.Thread(target=server.run)
-    thread.start()
-    while not server.started:
-        time.sleep(0.01)
-    response = requests.get("http://127.0.0.1:8000")
-
-    assert response.headers["server"] == "over-ridden" and response.headers["date"]
-
-    thread.join()
-
-
-def test_override_server_header_multiple_times():
-    config = Config(
-        app=app,
-        loop="asyncio",
-        limit_max_requests=1,
-        headers=[("Server", "over-ridden"), ("Server", "another-value")],
-    )
-    server = Server(config=config)
-    thread = threading.Thread(target=server.run)
-    thread.start()
-    while not server.started:
-        time.sleep(0.01)
-    response = requests.get("http://127.0.0.1:8000")
-
-    assert (
-        response.headers["server"] == "over-ridden, another-value"
-        and response.headers["date"]
-    )
-
-    thread.join()
-
-
-def test_add_additional_header():
-    config = Config(
-        app=app,
-        loop="asyncio",
-        limit_max_requests=1,
-        headers=[("X-Additional", "new-value")],
-    )
-    server = Server(config=config)
-    thread = threading.Thread(target=server.run)
-    thread.start()
-    while not server.started:
-        time.sleep(0.01)
-    response = requests.get("http://127.0.0.1:8000")
-
-    assert (
-        response.headers["x-additional"] == "new-value"
-        and response.headers["server"] == "uvicorn"
-        and response.headers["date"]
-    )
-
-    thread.join()
+@pytest.fixture()
+def server_fixture(event_loop, config_arg):
```

Comment:
This could even be rewritten as a `async def` fixture, since pytest-asyncio supports that.

Reply:
If I'm doing that I have a 
```
    @pytest.fixture
    async def server_fixture(event_loop, config_arg):
        server = Server(config=config_arg)
        cancel_handle = asyncio.ensure_future(server.serve(), loop=event_loop)
>       event_loop.run_until_complete(asyncio.sleep(0.1))

test_default_headers.py:19: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.asdf/installs/python/3.8.6/lib/python3.8/asyncio/base_events.py:592: in run_until_complete
    self._check_running()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_UnixSelectorEventLoop running=False closed=False debug=False>

    def _check_running(self):
        if self.is_running():
>           raise RuntimeError('This event loop is already running')
E           RuntimeError: This event loop is already running

../../../.asdf/installs/python/3.8.6/lib/python3.8/asyncio/base_events.py:552: RuntimeError
```

Reply:
Yes, I think you'd need to `await` directly :)

Reply:
Honestly though, I don't think a server fixture might be the best way to go. I see that we have to go thorough interesting hoops just to pass the config as a parameter.

What about a plain and simple `@asynccontextmanager`? Then we use it explicitly in tests via `async with run_server(config=...)`?